src/common/weighted_shuffle.h

```diff
@@ -0,0 +1,41 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+/*
+ * Ceph - scalable distributed file system
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation.  See file COPYING.
+ *
+ */
+
+#ifndef CEPH_WEIGHTED_SHUFFLE_H
+#define CEPH_WEIGHTED_SHUFFLE_H
+
+#include <vector>
+#include <random>
+#include <algorithm>
+
+template <class T, class R>
+auto weighted_shuffle(std::vector<T> data, std::vector<int> weights, R &&gen)
```

Comment:
i guess this implementation comes from https://stackoverflow.com/questions/50221136/c-weighted-stdshuffle . so it's licensed under CC-BY-SA unless the author claims otherwise.

but it removes the nice things from the original version:

1. we can do the shuffle in-place, and no need to pass the shuffled vector by value
2. we should not pass the weight by value.